docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,150 @@
+    sync'ing and how to build them will allow us to cover use cases like
+    springboot dev-tools.
+
+2. If this is a redesign, what are the drawbacks of the current implementation?
+  - This is not a redesign, but a new sync mode that may or may not be available
+    in the skaffold.yaml directly to users. It could simply be an internal API
+    that is usuable by builders like Jib.
+  - This is similar to `_smart_` described in [sync-improvements](sync-improvements.md)
+
+3. Is there any another workaround, and if so, what are its drawbacks?
+  - Currently one can create a docker file that only copies specific build
+    results into the container and relies on a local build to generate those
+    intermediate artifacts. This requires a user to trigger a first build
+    manually before docker kicks in to containerize the application. While
+    it may be possible to automate this (for example: gradle --continuous), it
+    is not an acceptable solution to require manual external processes for
+    a build to succeed.
+
+4. Mention related issues, if there are any.
+  - This is not trying to solve the problem of dealing with a multistage
+    dockerbuild. Intermediate build artifacts might still be possible to
+    determine, however that would require an extra mechanism to do a partial
+    docker build and sync files from a built container to a running container --
+    something we do not intend to cover here.
+
+#### Problems with current API/config
+The current `sync` system has the following problems:
+1. No way to trigger local external processes - for example, in jib, to build a
+   container, one would run `./gradlew jib`, but to update class files so the
+   system may sync them, one would only be required to run `./gradlew classes`.
+2. No way to tell the system to sync non build inputs. Skaffold is
+   normally only watching `.java` files for a build, in the sync case, we want
+   it to watch `.java` files, trigger a partial build, and sync `.class` files
+   so a remote server can pick up and reload the changes.
+
+## Design
+
+#### Hack it
+To get close to the functionality we want, without modifying skaffold at all, a
+Dockerfile which depends on java build outputs could be used, like:
+```
+FROM openjdk:8
+COPY build/dependencies/ /app/dependencies
+COPY build/classes/java/main/ /app/classes
+
+CMD ["java", "-cp", "/app/classes:/app/dependencies/*", "hello.Application"]
+```
+
+with a skaffold sync block that looks like:
+```
+sync:
+  manual:
+    - src: "build/classes/java/main/**/*.class"
+      dest: "/app/classes"
+      strip: "build/classes/java/main/"
+```
+
+A user's devloop then looks like this:
+
+1. run `./gradlew classes copyDependencies`
+2. run `skaffold dev`
+3. *make changes to some java file*
+4. run `./gradlew classes`
+5. *skaffold syncs files*
+
+which is far from ideal.
+
+#### Fix it?
+
+What we might want, is an API that can accept configuration that looks like:
+
+```
+sync: auto {
```

Comment:
We can drop `auto` and just add another mode `generated` along with existing `manual` and `infer`?
```suggestion+21
sync: 
  generated:
  - command: "./gradlew jib classes"
    inputs:
      - src/main/java
      - src/main/resources
    files:
      - src: "target/classes"
        dest: "/app/classes"
      - src: "target/resources"
        dest: "/app/resources"
  - command: "compile_c_stuff.sh"
    inputs:
    - asdf.c
    files:
     - src: "asdf.o"
       dest: "/libs/asdf.o"
  manual:
    - src: "local/computer/extrafiles"
      dest: "/extrafiles"
  includeTools: true
```

Reply:
I like this variant for the `yaml` syntax. As I understand, the existing `sync` specs would keep their semantics, but the `generated` is a new build hook to be triggered before sync. (Maybe `buildHooks` is also a more natural name than `generated`). In other words:

* if `generated` is present, execute the build hooks when changes matching `inputs` are detected
* if `generated` is absent -> current behavior.

Reply:
If a file is present in both `generated` and `manual` we would either have to give one preference or throw an error. Personally I'm leaning towards throwing an error, so that user's don't end up with unexpected sync stuff happening.